packages/kit/src/runtime/client/router.js

```diff
@@ -55,9 +57,20 @@ export class Router {
 
 		// create initial history entry, so we can return here
 		history.replaceState(history.state || {}, '', location.href);
+		// keeping track of the history index in order to prevent popstate navigation events if needed
+		this.current_history_index = 0;
+
+		/** @type {((url: URL) => void | boolean | Promise<void | boolean>)[]} */
+		this.on_before_navigate_callbacks = [];
 	}
 
 	init_listeners() {
+		if (history.state['sveltekit:index'] >= 0) {
+			this.current_history_index = history.state['sveltekit:index'];
+		} else {
+			history.replaceState({ ...history.state, 'sveltekit:index': 0 }, '', location.href);
+		}
+
```

Comment:
It might be better, to move this to [line 59](https://github.com/sveltejs/kit/blob/c86cac6553ce9ed3f65b0b62f7159422de1be88d/packages/kit/src/runtime/client/router.js#L59).

Reply:
![Screenshot_20220102_110125](https://user-images.githubusercontent.com/43446836/147871139-8e84733f-a7ad-4288-9e40-4844b65bb0df.png)

Is that how you mean it? Maybe we do not need the `if (history.state['sveltekit:index'] >= 0) {` at all if we assume the `sveltekit:index` is always 0 during initialization?

Reply:
No, that would overwrite the index on a reload.
The if-block is needed.

I just thought, it might be better, because it would be just one `replaceState` call on the initial page.

Reply:
Does it make sense to actually initialize the `this.current_history_index` with 0 in the constructor rather than in the if statement? And also is it ok to call `history.replaceState` conditionally rather than always? Maybe we could always call `history.replaceState` and make the `sveltekit:index` dynamic? Like that:

![Screenshot_20220102_112951](https://user-images.githubusercontent.com/43446836/147871770-f4720332-4988-425a-a4c1-1b0200f61f83.png)

Reply:
Hm maybe it does make sense to put it in a conditional statement, during reload the .replaceState will not do anything since we already have the state, right? So putting it inside a condition might be correct.

Reply:
The end-result would be the same.
Only that there would be always one `replaceState` call.

Reply:
> Hm maybe it does make sense to put it in a conditional statement, during reload the .replaceState will not do anything since we already have the state, right? So putting it inside a condition might be correct. But doesn't hurt I guess?

Yea, thats what meant

Reply:
Nice, I've pushed a new commit that will apply what we've discussed

Reply:
> Nice, I've pushed a new commit that will apply what we've discussed

Now the if statement makes no sense 😅

I think you meant it to be like:
```js
if (this.current_history_index === 0) { ... }
```

Or 
```js
if (history.state?.['sveltekit:index'] === undefined) { ... }
```

Reply:
correct, fixed

Reply:
oh wait not really

Reply:
now it is fixed